
#Function 1) Merging----
#This will merge all the .txt files in a specific sub folder
#In the Adult and Larvae subfolder
#Into one common data frame

#This function takes two input
# 1) list of file names in a directory
# 2) The "1_Combining_files.csv" that contains the path of 
#subfolders and the prefix that will be added in the column names
#The function returns the combined file.

merging <-function(list_of_file_names, f){
  message(list_of_file_names)
  myfile <- read.csv(list_of_file_names[1], sep="\t")
  df <- as.data.frame(cbind(myfile$gene_short_name, myfile$FPKM))
  names(df)[1] <- "Gene"
  temp <- strsplit(list_of_file_names[1], "_")[[1]]
  temp <- temp[length(temp)]
  temp <- strsplit(temp,".txt")
  names(df)[2] <- temp[[1]]
  for (i in 2:length(list_of_file_names)){
    myfile <- read.csv(list_of_file_names[i], sep="\t")
    temp_df <- as.data.frame(cbind(myfile$gene_short_name, myfile$FPKM))
    names(temp_df)[1] <- "Gene"
    temp <- strsplit(list_of_file_names[i], "_")[[1]]
    temp <- temp[length(temp)]
    temp <- strsplit(temp,".txt")
    names(temp_df)[2] <- temp[[1]]
    df <- merge(df, temp_df, by ="Gene")
  }
  col_names <- names(df)
  for (k in 2:length(col_names)){
    names(df)[k] <- paste0(f[j,2],col_names[k])
  }
  return(df)
}

#Function 2) ortholog----
#This function takes the files containing orthologs information as an input
#It will select only those genes that have orthologs and will merge all the data frames 
#generated by above function.
#The output will be a combined file, containing FPKM values for each experiment condition
ortholog <- function(ortho_file){
  f1 <- ortho_file
  f2 <- ortho_file
  #f <- read.csv("2_orthologs.txt", sep ="\t")
  f1$sim <- NULL
  merged_sec <- merge(f1,df1 , by.x="sec", by.y="Gene")
  merged_sec <- merge(merged_sec,df3 , by.x="sec", by.y="Gene")
  merged_sec <- merge(merged_sec,df5 , by.x="sec", by.y="Gene")
  merged_sec <- merge(merged_sec,df7 , by.x="sec", by.y="Gene")
  merged_sec$sec <- NULL
  
  #f <- read.csv("2_orthologs.txt", sep ="\t")
  f2$sec <- NULL
  merged_sim <- merge(f2,df2 , by.x="sim", by.y="Gene")
  merged_sim <- merge(merged_sim,df4 , by.x="sim", by.y="Gene")
  merged_sim <- merge(merged_sim,df6 , by.x="sim", by.y="Gene")
  merged_sim <- merge(merged_sim,df8 , by.x="sim", by.y="Gene")
  merged_sim$sim <- NULL
  
  f <- merge(merged_sim,merged_sec, by= "Gene")
  return(f)
}

#Function 3) PCA----
#This function will take the "Final_combined_FPKM.csv" as input
#and will generate a PCA plot for only Parents, only Hybrids and Combined dataset.
my_pca <- function(f, metadata){
  #pdf(file="Plots.pdf")
  p <- pca(f, metadata = metadata, removeVar = 0.1)
  horn <- parallelPCA(f)
  message(horn$n)
  elbow <- findElbowPoint(p$variance)
  message(elbow)
  plot(screeplot(p,components = getComponents(p, 1:20),vline = c(horn$n, elbow)) + 
    geom_label(aes(x = horn$n , y = 60,label = 'Horn\'s', vjust = -1, size = 8)) + 
    geom_label(aes(x = elbow, y = 30,label = 'Elbow method', vjust = -1, size = 8)))
  plot(pairsplot(p))
  plot(eigencorplot(p, metavars = c('Stage','Species','Previous_gen','Current_gen')))
  plot(eigencorplot(p,
               components = getComponents(p, 1:elbow),
               metavars = c('Stage','Species','Previous_gen','Current_gen'),
               col = c('white', 'cornsilk1', 'gold', 'forestgreen', 'darkgreen'),
               cexCorval = 1.2,
               fontCorval = 2,
               posLab = 'all',
               rotLabX = 45,
               scale = TRUE,
               main = bquote(Principal ~ component ~ Pearson ~ r^2 ~ correlates),
               plotRsquared = TRUE,
               corFUN = 'pearson',
               corUSE = 'pairwise.complete.obs',
               signifSymbols = c('****', '***', '**', '*', ''),
               signifCutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1)))
  plot(biplot(p))
  plot(biplot(p,
         #x = 'PC2', y = 'PC3',
         lab = paste0(p$metadata$Previous_gen,p$metadata$Current_gen),
         colby = 'Species',
         shape = 'Stage',
         hline = 0, vline = 0,
         legendPosition = 'right'))
  
  plot(biplot(p,
         x = 'PC2', y = 'PC3',
         lab = paste0(p$metadata$Previous_gen,p$metadata$Current_gen),
         colby = 'Species',
         shape = 'Stage',
         hline = 0, vline = 0,
         legendPosition = 'right'))
  plot(biplot(p,
         x = 'PC3', y = 'PC4',
         lab = paste0(p$metadata$Previous_gen,p$metadata$Current_gen),
         colby = 'Species',
         shape = 'Stage',
         hline = 0, vline = 0,
         legendPosition = 'right'))
  plot(plotloadings(p, labSize = 3))
  M<-cor(f)
  corrplot(M, type="upper", order="hclust",
           col=brewer.pal(n=8, name="PuOr"))
  corrplot(M, type="upper", order="hclust", tl.col="black", tl.srt=45)
  corrplot(M, method="pie")
  corrplot(M, method="number")
  f1 <- as.matrix(f)
  heatmap(f1)
  #dev.off()
}
#Function 4) Creating Factors and data file----
my_factors_and_data <- function(f){
  f<-t(f)
  f<-as.data.frame(f)
  categories<-rownames(f)
  categories<-as.data.frame(categories)
  categories$Previous<-NA
  categories$Current<-NA
  categories$Species<-NA
  categories$Stage<-NA
  categories$Generation<-NA
  rownames(categories)<-categories$categories
  for (val in rownames(categories)) {
    #print(val)
    #grep(pattern, string)
    #stringr::str_detect(string, pattern)
    if(str_detect(val,"Sec")){
      categories[val,"Species"]<-1
    } else{
      categories[val,"Species"]<-0
    }
  }
  
  for (val in rownames(categories)) {
    #print(val)
    #grep(pattern, string)
    #stringr::str_detect(string, pattern)
    if(str_detect(val,"Adult")){
      categories[val,"Stage"]<-0
    } else{
      categories[val,"Stage"]<-1
    }
  }
  for (val in rownames(categories)) {
    #print(val)
    #grep(pattern, string)
    #stringr::str_detect(string, pattern)
    if(str_detect(val,"Hybrid")){
      categories[val,"Generation"]<-1
    } else{
      categories[val,"Generation"]<-0
    }
  }
  
  for (val in rownames(categories)) {
    #print(val)
    #grep(pattern, string)
    #stringr::str_detect(string, pattern)
    if(str_detect(val,"CC")){
      categories[val,"Previous"]<-0
      categories[val,"Current"]<-0
    } else if(str_detect(val,"CO")){
      categories[val,"Previous"]<-0
      categories[val,"Current"]<-1
    } else if(str_detect(val,"OC")){
      categories[val,"Previous"]<-1
      categories[val,"Current"]<-0
    } else if(str_detect(val,"OO")){
      categories[val,"Previous"]<-1
      categories[val,"Current"]<-1
    }
  }
  mer<-merge(categories,f,by="row.names",sort = FALSE)
  mer$Row.names<-NULL
  write.table(mer,file="Factors_and_data.csv",
              sep=",", row.names = FALSE, quote = FALSE)
}
#Function 5) Linear Model----
my_linear_model <- function(mer,gene_start){
  i <- gene_start
  estimates <- NULL
  std_error <- NULL
  t_value <- NULL
  p_value <- NULL
  gene_name <- NULL
  other_parameters <- NULL
  model<-lm((log(mer[,i]+0.0001))~Previous*Current*Species*Stage*Generation,data=mer)
  estimates<-summary(model)$coefficients[,1]
  std_error<-summary(model)$coefficients[,2]
  t_value<-summary(model)$coefficients[,3]
  p_value<-summary(model)$coefficients[,4]
  gene_name<-names(mer)[i]
  estimates<-as.data.frame(t(estimates))
  std_error<-as.data.frame(t(std_error))
  t_value<-as.data.frame(t(t_value))
  p_value<-as.data.frame(t(p_value))
  gene_name<-as.data.frame(gene_name)
  estimates<-cbind(gene_name,estimates)
  std_error<-cbind(gene_name,std_error)
  t_value<-cbind(gene_name,t_value)
  p_value<-cbind(gene_name,p_value)
  max_residue<-max(model[["residuals"]])
  min_residue<-min(model[["residuals"]])
  median_residue<-median(model[["residuals"]])
  first_quantile<-quantile(model[["residuals"]])[2]
  third_quantile<-quantile(model[["residuals"]])[4]
  r_square<-summary(model)$r.squared
  adj_r_square<-summary(model)$adj.r.squared
  x<-as.data.frame(t(summary(model)$fstatistic))
  colnames(x)[1]<-"f_stats"
  model_pvalue<-pf(summary(model)$fstatistic[1],
                   summary(model)$fstatistic[2],
                   summary(model)$fstatistic[3],
                   lower.tail=FALSE)
  res_std_err<-sqrt(deviance(model)/df.residual(model))
  other_parameters<-cbind(gene_name,min_residue,first_quantile,
                          median_residue,third_quantile,max_residue,
                          res_std_err,r_square,adj_r_square,
                          model_pvalue,x)
  gene_start <- gene_start + 1
  for(i in gene_start:ncol(mer)){
    model<-lm((log(mer[,i]+0.0001))~Previous*Current*Species*Stage*Generation,data=mer)
    
    sample_estimates<-summary(model)$coefficients[,1]
    sample_std_error<-summary(model)$coefficients[,2]
    sample_t_value<-summary(model)$coefficients[,3]
    sample_p_value<-summary(model)$coefficients[,4]
    gene_name<-names(mer)[i]
    
    sample_estimates<-as.data.frame(t(sample_estimates))
    sample_std_error<-as.data.frame(t(sample_std_error))
    sample_t_value<-as.data.frame(t(sample_t_value))
    sample_p_value<-as.data.frame(t(sample_p_value))
    gene_name<-as.data.frame(gene_name)
    
    sample_estimates<-cbind(gene_name,sample_estimates)
    sample_std_error<-cbind(gene_name,sample_std_error)
    sample_t_value<-cbind(gene_name,sample_t_value)
    sample_p_value<-cbind(gene_name,sample_p_value)
    
    estimates<-rbind(estimates,sample_estimates)
    std_error<-rbind(std_error,sample_std_error)
    t_value<-rbind(t_value,sample_t_value)
    p_value<-rbind(p_value,sample_p_value)
    
    sample_max_residue<-max(model[["residuals"]])
    sample_min_residue<-min(model[["residuals"]])
    sample_median_residue<-median(model[["residuals"]])
    sample_first_quantile<-quantile(model[["residuals"]])[2]
    sample_third_quantile<-quantile(model[["residuals"]])[4]
    sample_r_square<-summary(model)$r.squared
    sample_adj_r_square<-summary(model)$adj.r.squared
    sample_x<-as.data.frame(t(summary(model)$fstatistic))
    colnames(sample_x)[1]<-"f_stats"
    sample_model_pvalue<-pf(summary(model)$fstatistic[1],summary(model)$fstatistic[2],summary(model)$fstatistic[3],lower.tail=FALSE)
    sample_res_std_err<-sqrt(deviance(model)/df.residual(model))
    sample_other_parameters<-cbind(gene_name,sample_min_residue,sample_first_quantile,sample_median_residue,
                                   sample_third_quantile,sample_max_residue,sample_res_std_err,sample_r_square,
                                   sample_adj_r_square,sample_model_pvalue,sample_x)
    colnames(sample_other_parameters)<-c("gene_name","min_residue","first_quantile","median_residue",
                                         "third_quantile","max_residue","res_std_err","r_square",
                                         "adj_r_square","model_pvalue","f_stats","numdf","dendf")
    other_parameters<-rbind(other_parameters,sample_other_parameters)
  }
  
  colnames(estimates)[2]<-"Intercept"
  colnames(t_value)[2]<-"Intercept"
  colnames(std_error)[2]<-"Intercept"
  colnames(p_value)[2]<-"Intercept"
  write.table(estimates, file="2_Estimates.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  write.table(std_error, file="3_Std_error.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  write.table(t_value, file="4_T_values.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  write.table(p_value, file="5_P_values.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  write.table(other_parameters, file="6_Other_parameters.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  other_parameters<-read.csv("6_Other_parameters.csv")
  temp_pvalue<-p_value
  temp_pvalue<-temp_pvalue[,-1]
  q_value<-matrix(p.adjust(as.vector(as.matrix(temp_pvalue)), method='fdr'),ncol=ncol(temp_pvalue))
  colnames(q_value)<-names(temp_pvalue)
  q_value<-as.data.frame(q_value)
  q_value<-cbind(p_value$gene_name,q_value)
  colnames(q_value)[1]<-"gene_name"
  write.table(q_value, file="7_QValues.csv",
              row.names = FALSE, quote = FALSE,
              sep=",")
  p_value_df<-data.frame(matrix(ncol = 1, nrow = 0))
  dta<-data.frame(matrix(ncol=1,nrow=9163))
  colnames(p_value_df)<-"Count"
  for(i in 2:ncol(p_value)){
    p_value_df[i-1,1]<-nrow(subset(p_value,p_value[,i]<0.05))
    gene_names <- p_value[p_value[,i]<0.05, 1]
    gene_names<-as.data.frame(gene_names)
    colnames(gene_names)<-names(p_value)[i]
    dta <- qpcR:::cbind.na(dta,gene_names)
  }
  temp<-as.data.frame(names(p_value))
  temp<-as.data.frame(temp[-1,])
  colnames(temp)<-"Condition"
  p_value_df<-cbind(temp,p_value_df)
  dta<-dta[,-1]
  write.table(p_value_df, file="8_PValue_count.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  write.table(dta, file="9_PValue_gene_names.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  q_value_df<-data.frame(matrix(ncol = 1, nrow = 0))
  dta<-data.frame(matrix(ncol=1,nrow=9163))
  colnames(q_value_df)<-"Count"
  for(i in 2:ncol(q_value)){
    q_value_df[i-1,1]<-nrow(subset(q_value,q_value[,i]<0.05))
    gene_names <- q_value[q_value[,i]<0.05, 1]
    gene_names<-as.data.frame(gene_names)
    colnames(gene_names)<-names(q_value)[i]
    dta <- qpcR:::cbind.na(dta,gene_names)
  }
  temp<-as.data.frame(names(q_value))
  temp<-as.data.frame(temp[-1,])
  colnames(temp)<-"Condition"
  q_value_df<-cbind(temp,q_value_df)
  dta<-dta[,-1]
  write.table(q_value_df, file="10_QValue_count.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  write.table(dta, file="11_QValue_gene_names.csv",
              row.names = FALSE, quote = FALSE, sep=",")
  #pdf(file="2_Linear_model/Combined/bar_graphs.pdf")
  plot(ggplot(data=p_value_df, aes(x=reorder(Condition,Count), y=Count)) +
    geom_bar(stat="identity", fill="steelblue")+
    geom_text(aes(label=Count), vjust=-0.3, size=3.5)+
    theme_minimal() + ggtitle("Significant P-Values")+ 
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    xlab("Categories"))
  
  plot(ggplot(data=q_value_df, aes(x=reorder(Condition,Count), y=Count)) +
    geom_bar(stat="identity", fill="steelblue")+
    geom_text(aes(label=Count), vjust=-0.3, size=3.5)+
    theme_minimal() + ggtitle("Significant Q-Values") +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    xlab("Categories"))
}
#Function 6)Cis/Trans interactions----
#The FPKM Values are plotted
#The categories (cis, trans, enhancing... etc) are assigned based on Qvalues


my_allele_interaction <- function(f,keyword,data){
  combinations <- c("CC", "CO", "OC", "OO")
  for (val in combinations){
    f1 <- f %>%
      dplyr::select(contains(val))
    parent <- f1 %>%
      dplyr::select(!contains("Hybrid"))
    hybrid <- f1 %>%
      dplyr::select(contains("Hybrid"))
    
    #P1->Adult Sim
    P1 <- as.data.frame(
      rowMeans(parent[,1:3], 
               na.rm = FALSE, dims = 1)
      )  
    colnames(P1) <- "P1"
    
    #P2->Adult Sec
    P2 <- as.data.frame(
      rowMeans(parent[,4:6], 
               na.rm = FALSE, dims = 1)
    )  
    colnames(P2) <- "P2"
    
    #A1-> Hybrid Sim
    A1 <- as.data.frame(
      rowMeans(hybrid[,1:3], 
               na.rm = FALSE, dims = 1)
    )  
    colnames(A1) <- "A1"
    
    #A2-> Hybrid Sec
    A2 <- as.data.frame(
      rowMeans(hybrid[,4:6], 
               na.rm = FALSE, dims = 1)
    )  
    colnames(A2) <- "A2"
    df <- cbind(P1,P2,A1,A2)
    df$'P1/P2' <- (df$P1+0.000001)/(df$P2+0.000001)
    df$'log2(P1/P2)' <- log2(df$`P1/P2`)
    df$'A1/A2' <- (df$A1+0.000001)/(df$A2+0.000001)
    df$'log2(A1/A2)' <- log2(df$`A1/A2`)
    df$Check <- paste0(keyword,"_",val)
    df <- cbind(df,data)
    
    #Assigning Category
    df$Category[(df$P_Species_Qvalues>0.05) & (df$H_Species_Qvalues>0.05) & (df$A_Species_Gen_Qvalues>0.05)] <- "Conserved"
    df$Category[(df$P_Species_Qvalues<0.05) & (df$H_Species_Qvalues<0.05) & (df$A_Species_Gen_Qvalues>0.05)] <- "Only_Cis"
    df$Category[(df$P_Species_Qvalues<0.05) & (df$H_Species_Qvalues>0.05) & (df$A_Species_Gen_Qvalues<0.05)] <- "Only_Trans"
    df$Category[(df$P_Species_Qvalues<0.05) & (df$H_Species_Qvalues<0.05) & (df$A_Species_Gen_Qvalues<0.05) & ((df$`log2(P1/P2)`/df$`log2(A1/A2)`)>1)] <- "Enhancing"
    df$Category[(df$P_Species_Qvalues<0.05) & (df$H_Species_Qvalues<0.05) & (df$A_Species_Gen_Qvalues<0.05) & ((df$`log2(P1/P2)`/df$`log2(A1/A2)`)<1)] <- "Compensating"
    #df$Category[(df$P_Species_Qvalues<0.05) & (df$H_Species_Qvalues<0.05) & (df$A_Species_Gen_Qvalues<0.05) & ((df$`log2(P1/P2)`/df$`log2(A1/A2)`)<1)] <- "Cis_by_Trans"
    df$Category[(df$P_Species_Qvalues>0.05) & (df$H_Species_Qvalues<0.05) & (df$A_Species_Gen_Qvalues<0.05)] <- "Compensating"
    #df$Category[(df$P_Species_Qvalues>0.05) & (df$H_Species_Qvalues<0.05) & (df$A_Species_Gen_Qvalues<0.05)] <- "Compensatory"
    df$Category[(df$P_Species_Qvalues<0.05) & (df$H_Species_Qvalues>0.05) & (df$A_Species_Gen_Qvalues>0.05)] <- "Ambiguous"
    df$Category[(df$P_Species_Qvalues>0.05) & (df$H_Species_Qvalues<0.05) & (df$A_Species_Gen_Qvalues>0.05)] <- "Ambiguous"
    df$Category[(df$P_Species_Qvalues>0.05) & (df$H_Species_Qvalues>0.05) & (df$A_Species_Gen_Qvalues<0.05)] <- "Ambiguous"

    df <- df[which(df$P1>0 & df$P2 > 0),]
    df <- df[which(df$P1>1 | df$P2 > 1),]
    df <- df[which(df$A1>0 & df$A2 > 0),]
    df <- df[which(df$A1>1 | df$A2 > 1),]
    
    a<-as.data.frame(table(df$Category))
    a <- subset(a, a$Var1 != "Ambiguous") #Comment this line if needed to include ambiguous
    lbls <- paste(a$Var1, a$Freq)
    lbls
    pie(a$Freq,labels = lbls, main=paste0(keyword,"_",val))
    
    #Conserved = Black
    Conserved<-df[which(df$Category=="Conserved"),]
    plot(Conserved$`log2(P1/P2)`, Conserved$`log2(A1/A2)`, pch = 19, col = "#000000", 
         xlim=c(-10,10), ylim=c(-10,10), main=paste0(keyword,"_",val), xlab = "log2(P1/P2)", ylab = "log2(A1/A2)")
    
    # Ambiguous<-df1[which(df1$Category=="Ambiguous"),]
    # points(Ambiguous$`log2(P1/P2)`,Ambiguous$`log2(A1/A2)`,pch = 19, col = "#8BE064", xlim=c(-10,10), ylim=c(-10,10))
    
    #Compensating = Green
    Compensating<-df[which(df$Category=="Compensating"),]
    points(Compensating$`log2(P1/P2)`,Compensating$`log2(A1/A2)`,pch = 19, col = "#8BE064", xlim=c(-10,10), ylim=c(-10,10))
    
    #Only Cis = Pink
    Only_Cis<-df[which(df$Category=="Only_Cis"),]
    points(Only_Cis$`log2(P1/P2)`,Only_Cis$`log2(A1/A2)`,pch = 19, col = "#FF00FF", xlim=c(-10,10), ylim=c(-10,10))
    
    #Only Trans = Purple
    Only_Trans<-df[which(df$Category=="Only_Trans"),]
    points(Only_Trans$`log2(P1/P2)`,Only_Trans$`log2(A1/A2)`,pch = 19, col = "#7137C8", xlim=c(-10,10), ylim=c(-10,10))
    
    #Enhancing = Peach
    Enhancing<-df[which(df$Category=="Enhancing"),]
    points(Enhancing$`log2(P1/P2)`,Enhancing$`log2(A1/A2)`,pch = 19, col = "#ffc1c1", xlim=c(-10,10), ylim=c(-10,10))
    
    abline(v=0)
    abline(h=0)
    abline(coef = c(0,1))
    abline(coef = c(0,-1))
    abline(v=0)
    abline(h=0)
    abline(coef = c(0,1))
    abline(coef = c(0,-1))
    df <- setDT(df, keep.rownames = "Gene")[]
    write.table(df, file=paste0(keyword,"_",val,".csv"),
                sep=",", quote = FALSE,
                row.names = FALSE)
     }
}
#Function 7) BarGraph----
# my_barGraph() <- function(){
#   f <- list.files(pattern = ".csv")
#   message(f)
#   df <- as.data.frame(matrix(ncol=2,nrow=0))
#   colnames(df) <- c("Check","Category")
#   for (i in f){
#     f1 <- read.csv(i)
#     Check <- f1$Check
#     Check <- as.data.frame(Check)
#     Category <- f1$Category
#     Category <- as.data.frame(Category)
#     temp <- cbind(Check,Category)
#     df <- rbind(df,temp)
#     write.table(df, file="Bar_graph1.csv", sep=",",
#                 quote= FALSE, row.names = FALSE)
#   }
#   data <- as.data.frame(table(df))
#   
#   for (i in unique(data$Category)){
#     temp <- subset(data, data$Category == i)
#     my_bar <- barplot(height=temp$Freq,
#             names=temp$Check,
#             col="#69b3a2",
#             main=i)
#     text(my_bar, temp$Check,paste(temp$Freq))
#   }
# }

#Function 8) Odd_ratios----
my_oddRatios <- function(f1 , file_list){
  
  for (i in file_list){
    f <- read.csv(i)
    f <- as.data.frame(cbind(f$Gene, f$Check, f$Category))
    names(f) <- c("Gene", "Check", "Category")
    df <- data.frame(matrix(ncol = ncol(f1), nrow = 6))
    names(df) <- names(f1)
    row.names(df) <- c("Size", "Conserved", "Only_Cis",
                       "Only_Trans", "Enhancing", "Compensating")
    colnames(df)[1] <- "Genome"
    col_names <- colnames(f1)
    for (j in 1:length(col_names)){
      x <- f1[,j]
      x <- na.omit(x)
      x <- as.data.frame(x)
      colnames(x) <- col_names[j]
      df[1,j] <- nrow(x)
      mer <- merge(f, x, by.x="Gene", by.y= names(x))
      df[2,j] <- nrow(mer[which(mer$Category == "Conserved"),])
      df[3,j] <- nrow(mer[which(mer$Category == "Only_Cis"),])
      df[4,j] <- nrow(mer[which(mer$Category == "Only_Trans"),])
      df[5,j] <- nrow(mer[which(mer$Category == "Enhancing"),])
      df[6,j] <- nrow(mer[which(mer$Category == "Compensating"),])
    }
    
    df_pvalue <- df
    df_odd_ratios <- df
    df_pvalue <- df_pvalue[-1,-1]
    df_odd_ratios <- df_odd_ratios[-1,-1]
    
    for (m in 2:nrow(df)){
      for (n in 2:ncol(df)){
        mini <- df[c(1,m), c(1,n)]
        test_results <- fisher.test(mini)
        df_pvalue[m-1,n-1] <- test_results[["p.value"]]
        df_odd_ratios[m-1,n-1] <- test_results[["estimate"]]
      }
    }
    
    #Log2 transforming odd ratios
    for (m in 1:nrow(df_odd_ratios)){
      for (n in 1:ncol(df_odd_ratios)){
        df_odd_ratios[m,n] <-log2(df_odd_ratios[m,n] + 0.000000001)
      }
    }
    
    #Converting Non-significant values to zero
    for (m in 1:nrow(df_odd_ratios)){
      for (n in 1:ncol(df_odd_ratios)){
        if (df_pvalue[m,n] > 0.05){
          df_odd_ratios[m,n] <- 0
        }
      }
    }
    
    #par(mar = c(5, 6, 4, 2) + 0.5)
    #heatmap(as.matrix(df_odd_ratios))
    
    df_odd_ratios_2 <- df_odd_ratios
    df_odd_ratios_2 <- ifelse(df_odd_ratios > 0, 1, 
                              ifelse(df_odd_ratios < 0, -1, 0))
    
    paletteLength <- 100
    #my_color <- colorRampPalette(c("yellow", "white", "blue"))(paletteLength)
    my_color <- colorRampPalette(c("blue", "white", "red"))(paletteLength)
    
    # myBreaks <- c(seq(min(df_odd_ratios), 0,
    #                   length.out = ceiling(paletteLength/2)+1),
    #               seq(max(df_odd_ratios)/paletteLength, max(df_odd_ratios),
    #                   length.out = floor(paletteLength/2)+50))
    
    #1 original
    pheatmap(as.matrix(df_odd_ratios_2), main = f[1,2])#, breaks = myBreaks)
    #2 original without clustering
    pheatmap(as.matrix(df_odd_ratios_2), main = f[1,2],
             cluster_rows = FALSE, cluster_cols = FALSE)#, breaks = myBreaks)
    
    df_odd_ratios <- setDT(df_odd_ratios, keep.rownames = "Category")[]
    write.table(df_odd_ratios, file=paste0("Odd_ratios_",f[1,2],".csv"),
                sep=",", quote = FALSE, 
                row.names = FALSE)
    
    df_pvalue <- setDT(df_pvalue, keep.rownames = "Category")[]
    write.table(df_pvalue, file=paste0("Fisher_Test_Pvalue_",f[1,2],".csv"),
                sep=",", quote = FALSE, 
                row.names = FALSE)
}#first for
  df <- read.csv("Odd_ratios_Adult_CC.csv")
  df$Fname <- NA
  df <- subset(df, FALSE)
  pattern <- "Odd_ratios_(Adult|Larvae).*\\.csv$"
  final_heatmap_list <- list.files(pattern = pattern)
  final_heatmap_list
  for (k in final_heatmap_list){
    f3 <- read.csv(k)
    temp <- strsplit(k, "_")[[1]]
    temp <- paste0(temp[3],"_", temp[4])
    temp <- strsplit(temp,".csv")[[1]]
    f3$Fname <- temp
    df <- rbind(df,f3)
  }
  df$Rname <- paste0(df$Category,"_",df$Fname)
  df$Category <- NULL
  df$Fname <- NULL
  rownames(df) <- df$Rname
  df$Rname <- NULL
  
  df <- ifelse(df > 0, 1,
               ifelse(df < 0, -1, 0))
  
  #1 original
  pheatmap(as.matrix(df), main = "All conditions")#, breaks = myBreaks)
  #2 original without clustering
  pheatmap(as.matrix(df), main = "All conditions",
           cluster_rows = FALSE, cluster_cols = FALSE)#, breaks = myBreaks)
  
} #function

#Representative heatmap----
#This function creates a representative heatmap
representative_map <- function(list_of_files, keyword){
  f <- read.csv(list_of_files[1], row.names = 1)
  for (n in 2:length(list_of_files)){
    f1 <- read.csv(list_of_files[n], row.names = 1)
    for (i in 1:nrow(f1)){
      for (j in 1:ncol(f1)){
        if (f1[i,j] > 0){
          f[i,j] <- max(f[i,j], f1[i,j])
        } else if (f1[i,j] < 0){
          f[i,j] <- min(f[i,j], f1[i,j])
        }
      }
    }
  }
  f <- ifelse(f > 0, 1,
               ifelse(f < 0, -1, 0))
  paletteLength <- 100
  #my_color <- colorRampPalette(c("yellow", "white", "blue"))(paletteLength)
  my_color <- colorRampPalette(c("blue", "white", "red"))(paletteLength)
  pheatmap(as.matrix(f), color = my_color, main = keyword)
  pheatmap(as.matrix(f), color = my_color, main = keyword,
           cluster_rows = TRUE, cluster_cols = FALSE)
}
  


